traffic_ops/traffic_ops_golang/role/roles.go

```diff
@@ -273,3 +301,369 @@ RETURNING id, last_updated`
+	} else if !ok {
+		api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no such role"), nil)
+		return
+	}
+
+	assignedUsers := 0
+	if err := inf.Tx.Get(&assignedUsers, "SELECT COUNT(id) FROM tm_user WHERE role=$1", roleID); err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("role delete counting assigned users: "+err.Error()))
+		return
+	} else if assignedUsers != 0 {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, fmt.Errorf("can not delete a role with %d assigned users", assignedUsers), nil)
+		return
+	}
+
+	rows, err := tx.Query(deleteRoleQuery(), roleID)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("deleting role: "+err.Error()))
+		return
+	}
+	rows.Close()
+	userErr, sysErr, errCode = deleteRoleCapabilityAssociations(inf.Tx, roleID)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	alerts := tc.CreateAlerts(tc.SuccessLevel, "role was deleted.")
+	api.WriteAlerts(w, r, http.StatusOK, alerts)
+	changeLogMsg := fmt.Sprintf("ROLE: %s, ID: %d, ACTION: Deleted Role", roleName, roleID)
+	api.CreateChangeLogRawTx(api.ApiChange, changeLogMsg, inf.User, tx)
+}
+
+// Create will create a new role based on the struct supplied.
+func Create(w http.ResponseWriter, r *http.Request) {
+	var roleID int
+	var roleName string
+	var roleDesc string
+	var privLevel int
+	var roleCapabilities *[]string
+	var roleV4 tc.RoleV4
+
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	tx := inf.Tx.Tx
+	if err := json.NewDecoder(r.Body).Decode(&roleV4); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	if err := roleV4.Validate(); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+	current, err := auth.GetCurrentUser(r.Context())
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+		return
+	}
+	missing := current.MissingPermissions(roleV4.Permissions...)
+	if len(missing) != 0 {
+		api.HandleErr(w, r, tx, http.StatusForbidden, fmt.Errorf("cannot request more than assigned permissions, current user needs %s permissions", strings.Join(missing, ",")), nil)
+		return
+	}
+	roleName = roleV4.Name
+	roleDesc = roleV4.Description
+	privLevel = inf.User.PrivLevel
+	roleCapabilities = &roleV4.Permissions
+
+	rows, err := tx.Query(createQuery(), roleName, roleDesc, privLevel)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("creating role: %w", err))
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var throwaway interface{}
+		if err := rows.Scan(&roleID, &throwaway); err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("role create: scanning role ID: %w", err))
+			return
+		}
```

Comment:
`createQuery` is only called in one place, because legacy handlers are using `insertQuery` instead. Therefore, there's no reason to have a throwaway here; the query can just return only and exactly what it needs to - which should include the value for the response's `lastUpdated` field, by the way.